[SPARK-42199][SQL] Fix issues around Dataset.groupByKey

EnricoMi opened the pull request:
### What changes were proposed in this pull request?
Introduces `ScopedExpression`, which allows to resolve an expression against a set of attributes. This is used by `Dataset.groupByKey.agg`, `Dataset.groupByKey.flatMapSortedGroups`, and `Dataset.groupByKey.cogroupSorted` to fix surprising error messages and to make `Dataset.groupByKey.mapValues.agg` work at all.

### Why are the changes needed?
Calling `ds.groupByKey(func: V => K)` creates columns to store the key value. These columns may conflict with columns that already exist in `ds`. Function `Dataset.groupByKey.agg` accounts for this with a very specific rule, which has some surprising weaknesses:

```Scala
spark.range(1)
  // groupByKey adds column 'value'
  .groupByKey(id => id)
  // which cannot be referenced, though it is suggested
  .agg(count("value"))
```
```
org.apache.spark.sql.AnalysisException: Column 'value' does not exist.
Did you mean one of the following? [value, id];
```

An existing 'value' column can be referenced:

```Scala
// dataset with column 'value'
spark.range(1).select($"id".as("value")).as[Long]
  // groupByKey adds another column 'value'
  .groupByKey(id => id)
  // agg accounts for the extra column and excludes it when resolving 'value'
  .agg(count("value"))
  .show()
```
```
+---+------------+
|key|count(value)|
+---+------------+
|  0|           1|
+---+------------+
```

While column suggestion shows both 'value' columns:

```Scala
spark.range(1).select($"id".as("value")).as[Long]
  .groupByKey(id => id)
  .agg(count("unknown"))
```
```
org.apache.spark.sql.AnalysisException: Column 'unknown' does not exist.
Did you mean one of the following? [value, value]
```

However, `mapValues` introduces another 'value' column, which should be referencable, but it breaks the exclusion introduced by `agg`:

```Scala
spark.range(1)
  // groupByKey adds column 'value'
  .groupByKey(id => id)
  // adds another 'value' column
  .mapValues(value => value)
  // which cannot be referenced in agg
  .agg(count("value"))
```
```
org.apache.spark.sql.AnalysisException: Reference 'value' is ambiguous, could be: value, value.
```

### Does this PR introduce _any_ user-facing change?
Fixes suggestions in above error messages from
```
... Did you mean one of the following? [value, value]
... Did you mean one of the following? [value, id];
```
to
```
... Did you mean one of the following? [value]
... Did you mean one of the following? [id];
```

Allows for `Dataset.groupByKey.mapValues.agg`.

### How was this patch tested?
Tests in `DatasetSuite`.

EnricoMi committed:
Introduce ScopedExpression to control scope of expression resolution

EnricoMi commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -409,6 +409,20 @@ case class PrettyAttribute(
+/**
+ * An expression that has to be resolved against a scope of resolved attributes.
+ */
+case class ScopedExpression(child: Expression, scope: AttributeSeq)
```
One unsolved problem here is when attributes in `scope` get rewritten elsewhere in the plan, they should be rewritten in `scope` as well. For instance, `DeduplicateRelations` rule.

https://github.com/G-Research/spark/compare/branch-append-columns-improvements...G-Research:spark:branch-cogroup-self-deduplication-bug-after-append-column#diff-b98c99535d2b28cb47774860d500030e732c244c55b1ac05aead5d1cf1e7a602R1015-R1025

EnricoMi commented:
@cloud-fan @dongjoon-hyun are you interested in fixing these inconsistencies?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -3286,6 +3265,25 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
+  object ResolveScopedExpression extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveExpressions {
+      case se: ScopedExpression if se.resolved => se.child
+      case se @ ScopedExpression(expr, attributes) =>
```
one thing is we need to make sure `attributes` is a subset of the children output.

EnricoMi replied:
> we need to make sure `attributes` is a subset of the children output.

I think this only holds for `AttributeReference`, because `UnresolvedAttribute` that cannot be resolved against children output will cause an Analyzer error.

cloud-fan replied:
a lambda can match the first child of `CoGroup` and apply special logic if it's `AppendColumns`, another lambda can match the second child of `CoGroup`.

EnricoMi replied:
> apply special logic if it's AppendColumns,

Sounds like you still want to bypass the children if it is a `AppendColumns`, as we used to do in (and few other places):
```
object ResolveReferencesInAggregate extends SQLConfHelper
  with ColumnResolutionHelper with AliasHelper {
  def apply(a: Aggregate): Aggregate = {
    val planForResolve = a.child match {
      // SPARK-25942: Resolves aggregate expressions with `AppendColumns`'s children, instead of
      // `AppendColumns`, because `AppendColumns`'s serializer might produce conflict attribute
      // names leading to ambiguous references exception.
      case appendColumns: AppendColumns => appendColumns
      case _ => a
    }
...
```

We cannot know whether the child of a `MapGroups` / `CoGroup` is an `AppendColumns` that adds the key column (which we want to bypass or ignore), or an `AppendColumns` that comes from `mapValues` (which we do not want to ignore). See the _"However, mapValues introduces another 'value' column, ..."_ section in the PR description above.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala:
```diff
@@ -21,7 +21,7 @@ import scala.collection.JavaConverters._
 
```
`def mapValues` in this file is not updated, shouldn't it use `ScopedExpression`?

EnricoMi replied:
No, as it is fine to reference the `value` column provided by `mapValue`'s `AppendColumns`. That is what we want. `ScopedExpression` hides unwanted columns, here we do not want to hide `value`.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -410,6 +410,20 @@ case class PrettyAttribute(
+/**
+ * An expression that has to be resolved against a scope of resolved attributes.
+ */
+case class ScopedExpression(child: Expression, scope: AttributeSeq)
```
shall we put `Seq[Attribute]` and include it as children?

EnricoMi replied:
> shall we put Seq[Attribute]

yes, done:
```diff
-case class ScopedExpression(child: Expression, scope: AttributeSeq)
+case class ScopedExpression(child: Expression, scope: Seq[Attribute])
```

> include it as children?

as in
```diff
 case class ScopedExpression(child: Expression, scope: Seq[Attribute])
-  extends UnaryExpression with Unevaluable {
+  extends Expression with Unevaluable {
   override def dataType: DataType = child.dataType
   override def nullable: Boolean = child.nullable
   override def prettyName: String = "scoped"
 
   override def sql: String = s"$prettyName(${child.sql}, $scope)"
 
-  override def withNewChildInternal(newChild: Expression): ScopedExpression = copy(child = newChild)
+  override def withNewChildren(newChildren: Seq[Expression]): ScopedExpression = {
+    val scope = newChildren.tail
+    assert(scope.forall(_.isInstanceOf[Attribute]), "Scope children have to be attributes")
+    copy(child = newChildren.head, scope = scope.map(_.asInstanceOf[Attribute]))
+  }
 }
```

EnricoMi commented:
I have moved the logic that adds the `ScopedExpression` into a rule because some code paths (especially `connect`) call into the case class, which cannot rewrite the given sort argument(s). Alternative would be to duplicate that logic.

All tests green.

EnricoMi commented:
@cloud-fan all tests green and I think all comments are addressed

cloud-fan commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:
```diff
@@ -2422,6 +2422,9 @@ class SparkConnectPlanner(
+      // SPARK-42199: resolve these aggregate expressions only against dataAttributes
+      // this is to hide key column from expression resolution
+      .map(ScopedExpression(_, ds.dataAttributes))
       .map(toNamedExpression)
```
does this change the final expression name?

EnricoMi replied:
Not sure if it does, not sure how to test this either. Should this guarantee the scoped expression to have the same expression name: b7cf27d068ce274ba100e7ac990623b70917d0b7

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -3468,6 +3448,52 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
+    private def scopeOrder(scope: Seq[Attribute])(sortOrder: SortOrder): SortOrder = {
+      sortOrder match {
+        case so if so.child.isInstanceOf[ScopedExpression] => so
+        case so => so.copy(child = ScopedExpression(so.child, scope))
```
`SortOrder` has `sameOrderExpressions`, shall we scope them as well, or assert they are empty here?

EnricoMi replied:
Added in 894368e1432.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -302,6 +308,16 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
         new ResolveReferencesInAggregate(catalogManager).checkUnresolvedGroupByAll(operator)
 
+        getAllExpressions(operator).foreach(_.foreach {
+          case se: ScopedExpression if !se.resolved =>
```
can we also fail if `ScopedExpression` is still present? I think the analyzer should have removed it if its child is resolved, but just in case if bugs.

EnricoMi replied:
Added in b0510ff7bfd.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:
```diff
@@ -415,6 +415,25 @@ case class PrettyAttribute(
+/**
+ * An expression that has to be resolved against a scope of resolved attributes.
+ */
+case class ScopedExpression(expr: Expression, scope: Seq[Attribute])
```
This is not a named expression. shall we move it to a new file, with the two new analyzer rules?

EnricoMi replied:
As in bb1468ad876ccc9c77c3af25945c35301a5f9114?

cloud-fan commented:
@EnricoMi sorry this PR is lost track. Have you addressed all the review comments?

EnricoMi commented:
> Have you addressed all the review comments?

@cloud-fan all comments addressed, see above.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -519,6 +521,7 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
         case e if !e.resolved => u
         case g: Generator => MultiAlias(g, Nil)
         case c @ Cast(ne: NamedExpression, _, _, _) => Alias(c, ne.name)()
+        case se @ ScopedExpression(ne: NamedExpression, _) => Alias(se, ne.name)()
```
just in case, shall we also match `ScopedExpression(ExtractValue...)` and `ScopedExpression(Literal...)`?

EnricoMi replied:
What about
```suggestion
        case se @ ScopedExpression(e: Expression, _) =>
          resolve(UnresolvedAlias(e, optGenAliasFunc)) match {
            case ne: NamedExpression => Alias(se, ne.name)()
            case _ => se
          }
```

cloud-fan replied:
LGTM

EnricoMi replied:
Fixed in 9a360dd6465.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -310,6 +316,22 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
 
         var stagedError: Option[() => Unit] = None
+
+        getAllExpressions(operator).foreach(_.foreach {
```
there is another `getAllExpressions(operator).foreach(_.foreachUp {` below. Can we handle `ScopedExpression` there?

EnricoMi replied:
moved into next block of checks

EnricoMi replied:
Moving the checks into the next block makes the old error messages (wrong suggested columns) resurface:
- https://github.com/G-Research/spark/actions/runs/10684568288/job/29615369891#step:10:31013
- https://github.com/G-Research/spark/actions/runs/10684568288/job/29615369891#step:10:31076
- https://github.com/G-Research/spark/actions/runs/10684568288/job/29615369891#step:10:31139
- https://github.com/G-Research/spark/actions/runs/10684568288/job/29615369891#step:10:31202

Even if I move the checks to the top of the block. So they have to be checked in isolation of the other checks.

EnricoMi replied:
reverted 66d3471efea7dad49e915054d7480521689d9b3f

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/scopes.scala:
```diff
@@ -0,0 +1,93 @@
+/**
+ * An expression that has to be resolved against a scope of resolved attributes.
+ */
+case class ScopedExpression(expr: Expression, scope: Seq[Attribute])
```
I'm a bit worried about rule order issue here. How can we guarantee that we resolve `expr` with the scoped attributes? `expr` is a normal child of `ScopedExpression` and any rule can resolve it?

EnricoMi replied:
If we removed `exprs` from `override def children`, then no rule can resolve `exprs`, right?

EnricoMi replied:
For instance, for a groupBy.agg, the scope is the data attributes of the groupBy, which are not required to be resolved at that stage.

EnricoMi committed:
Reuse ScopedExpression child name

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/scopes.scala:
```diff
@@ -0,0 +1,93 @@
+    !sortOrder.child.isInstanceOf[ScopedExpression]
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+    // SPARK-42199: sort order of MapGroups must be scoped to their dataAttributes
```
We need to revisit the design. We need a way to make sure the attributes inside `ScopedExpression` are resolved by the scoping attributes, but the current design can't guarantee it, as any rule can resolve attributes inside `ScopedExpression` without respecting the scoping attributes.

EnricoMi replied:
Any suggestion? I lack the knowledge of the codebase to suggest a better design. Who can help here?

cloud-fan replied:
I think we should look into https://github.com/apache/spark/pull/39754/files#r1747046637 and figure out if there are other rules that need to resolve `ScopedExpression`.

cloud-fan commented:
seems there are compilation errors.

EnricoMi committed:
Make resolveExpressionByPlanOutput and …ByPlanChildren not resolve scoped expression

EnricoMi commented:
I have rebased with master and resolved conflicts.

I have also added this attempt to avoid resolving the scoped expressionby logical plan output / children: c6dab522e057cc2c339b7d0fcb215555ffaefbc4

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:
```diff
@@ -196,6 +196,8 @@ trait ColumnResolutionHelper extends Logging with DataTypeErrorsBase {
             u.copy(child = newChild)
           }
 
+        case s : ScopedExpression => s.mapScope(innerResolve(_, isTopLevel = false))
```
so we may resolve the `UnresolvedAttribute` inside `ScopedExpression` using the `resolveColumnByName` (which is usually the child plan output), instead of the scoped attributes?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/scopes.scala:
```diff
@@ -0,0 +1,102 @@
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+    // SPARK-42199: sort order of MapGroups must be scoped to their dataAttributes
+    case mg: MapGroups if mg.dataOrder.exists(isNotScoped) =>
+      mg.copy(dataOrder = mg.dataOrder.map(scopeOrder(mg.dataAttributes)))
```
at this time, `mg.dataAttributes` maybe unresolved, right?

EnricoMi replied:
right

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/scopes.scala:
```diff
@@ -0,0 +1,102 @@
+      mg.copy(dataOrder = mg.dataOrder.map(scopeOrder(mg.dataAttributes)))
+
+    // SPARK-42199: sort order of CoGroups must be scoped to their respective dataAttributes
+    case cg: CoGroup if Seq(cg.leftOrder, cg.rightOrder).exists(_.exists(isNotScoped)) =>
```
I'm wondering if we should add `ScopedExpression` in the Dataset API implementations instead of a new analyzer rule. Which is better for Spark Connect?

cloud-fan commented:
I'm thinking about this again but still not comfortable with this `ScopedExpression` approach, given it relies on rule execution order.

I have a new idea: how about we add a special annotation to the columns produced by `groupByKey`? It can be a special key in `AttributeRefererence#metadata`, and column resolution will skip it when resolving column names.

EnricoMi commented:
I'll try to sketch a fix using AttributeReference metadata.
